### DISCUSSION TITLE
ERROR\t -> :branch: \"production\" === :branch does not return true

### DISCUSSION THREAD

**Discussion by cozsmin** [puppetlabs/community#63] [Puppet Enterprise]
Hello,

First of all , I have doe a search on this forum and the error above does not show any similar posts/threads .

I am trying to setup a minimal Code manager as described in https://www.puppet.com/docs/pe/2023.6/code_mgr_config#code_mgr_enable

The  OS is ubuntu 22.04 , kernel 6.1.0-1034-oem , puppet package versions are :
e-puppet-enterprise-release       2023.6.0.0-1jammy
pe-puppetdb                                 2023.4.0.76-1jammy
pe-puppetdb-termini                   2023.4.0.76-1jammy
pe-puppetserver	                       2023.6.0.24-1jammy
puppet-agent                                8.4.0-1jammy

So far I have generated the /etc/puppetlabs/puppetserver/ssh/id-control_repo.ed25519 and /etc/puppetlabs/puppetserver/ssh/id-control_repo.ed25519.pub files.
Ran : puppet infrastructure configure

For the puppet_enterprise::profile::master class set 'code_manager_auto_configure' to 'true'.
'r10k_remote' set to 'git@github.com:cozsmin/puppet_control-repo00.git'
'r10k_private_key' = '/etc/puppetlabs/puppetserver/ssh/id-control_repo.ed25519'
'r10k_known_hosts' = [{"name":"github.com","type":"ssh-ed25519","key":"AAAAC3NzaC1lZDI1NTE5AAAAIOMqqnkVzrm0SdG6UOoqKLsabgH5C9okWi0dh2l9GKJl"}]

Created a puppet user and added him to the 'code deployers' group

Executed : puppet-access login --lifetime 180d

The 'Puppetfile' and 'environment.conf'  git repo files are :
Puppetfile :
mod 'mod00', :latest
&nbsp;&nbsp;    :git => 'git@github.com:cozsmin/puppet-control_repo.git',
&nbsp;&nbsp;    :branch => 'production'

environment.conf:
modulepath=site-modules:modules:$basemodulepath

The repo has 2 branches 'production' and 'deployment00' , where 'deployment00' is a copy of 'production' , and 'production' is the main one.

As per the puppet doc I executed the command 'puppet-code deploy deployment00 --wait' , and the output is :

Found 0 environments.
Found 1 environments.
[
&nbsp;  {
&nbsp;&nbsp;    "environment": "deployment00",
&nbsp;&nbsp;    "error": {
&nbsp;&nbsp;&nbsp;      "details": {
&nbsp;&nbsp;&nbsp;&nbsp;        "corrected-name": "deployment00"
&nbsp;&nbsp;&nbsp;      },
&nbsp;&nbsp;&nbsp;      "kind": "puppetlabs.code-manager/deploy-failure",
&nbsp;&nbsp;&nbsp;      "msg": "Errors while deploying environment 'deployment00' (exit code: 1):\nERROR\t -> :branch: \"production\" === :branch does not return true\n"
&nbsp;&nbsp;    },
&nbsp;&nbsp;    "id": 27,
&nbsp;&nbsp;    "status": "failed"
&nbsp;  }
]

I have no idea why this shows because git shows me i have the 2 branches :

root@puppet-enterprise:~/GIT/puppet_control-repo00# git branch
&nbsp;&nbsp;  deployment00
&nbsp;&nbsp;* production
root@puppet-enterprise:

Does anybody know ?

**Comment by binford2k** [comment]
Had you pushed your branches to https://github.com/cozsmin/puppet-control_repo when you saw this error? I see you now have only a `production` branch. Are you still seeing that error?

In any case, it looks like you're configuring r10k to look for a *module* where you have a *control repo*. 

What you should do is configure r10k with your control repo as the remote. Then you put your `environment.conf` and your `Puppetfile` In that control repo. R10k will read it *from GitHub* and deploy from there. Your local clone is only used for development.

**Reply by Sharpie** [reply]
The `mod` statement in that `Puppetfile` isn't valid syntax and produces a different error when evaluated in a test environment:

```
# mkdir tst && cd tst

# cat <<'EOF' >Puppetfile
mod 'mod00', :latest
   :git => 'git@github.com:cozsmin/puppet-control_repo.git',
   :branch => 'production'
EOF

# /opt/puppetlabs/puppet/bin/r10k puppetfile install
ERROR	 -> Failed to evaluate /root/tst/Puppetfile
Original exception:
/root/tst/Puppetfile:3: syntax error, unexpected string literal, expecting local variable or method
   :branch => 'production'
```

Could you double-check the `Puppetfile` content at `/etc/puppetlabs/code-staging/environments/deployment00/Puppetfile` and re-post it if different?

**Comment by cozsmin** [comment]
Hello and thank you for the comments.

I have put a wrong "Puppetfile" in my 1'st post , here is the real one "/etc/puppetlabs/code-staging/environments/deployment00/Puppetfile" : 

```
mod 'mod00',
    :git => 'git@github.com:cozsmin/puppet_control-repo00.git'
    :branch => 'production'
```
The difference is the ":git" is different , I have 2 repos , and I put the wrong one .

Sharpie , this is what I have done : 
```

root@puppet-enterprise:~/tmp# cat Puppetfile 
mod 'mod00',
    :git => 'git@github.com:cozsmin/puppet_control-repo00.git'
    :branch => 'production'
root@puppet-enterprise:~/tmp#

root@puppet-enterprise:~/tmp# /opt/puppetlabs/puppet/bin/r10k puppetfile install
ERROR    -> :branch: "production" === :branch does not return true
root@puppet-enterprise:~/tmp#
```

Still the same error.

These are the branches in the correct repo :

```
root@puppet-enterprise:~/GIT/puppet_control-repo00#  git branch
  deployment00
* production
root@puppet-enterprise:~/GIT/puppet_control-repo00# 
```

**Reply by cozsmin** [reply]
Ok , problem solved.

The solution came from https://stackoverflow.com/questions/70441482/puppet-code-manager-setup-issue-with-bitbucket.

I had to run 
`runuser -u pe-puppet -- /opt/puppetlabs/puppet/bin/r10k -c /opt/puppetlabs/server/data/code-manager/r10k.yaml deploy environment production --puppetfile
`
and
`runuser -u pe-puppet -- /opt/puppetlabs/puppet/bin/r10k -c /opt/puppetlabs/server/data/code-manager/r10k.yaml deploy environment production --puppetfile`

Now "puppet-code deploy deployment00 --wait" is working good .

Thank you all for the responses ....

**Comment by akomakom** [comment]
I am having the same issue but with `https` git repos, so it is not an SSH key issue. The solution above does not appear to help.  
I opened #91 for that.
